src/main/kotlin/org/rust/ide/intentions/createFromUsage/CreateStructIntention.kt

```diff
@@ -0,0 +1,96 @@
+    override fun getFamilyName() = "Create struct"
+
+    class Context(val name: String, val literalElement: RsStructLiteral, val target: RsMod) {}
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val path = element.parentOfType<RsPath>()
+        val structLiteral = path?.parentOfType<RsStructLiteral>()
+        if (structLiteral != null) {
+            if (structLiteral.path != path) return null
+            if (path.resolveStatus != PathResolveStatus.UNRESOLVED) return null
+
+            val target = getTargetModForStruct(path) ?: return null
+            val name = path.referenceName ?: return null
+
+            text = "Create struct `$name`"
+            return Context(name, structLiteral, target)
+        }
+        return null
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val struct = buildStruct(project, ctx) ?: return
+        val function = ctx.literalElement.parentOfType<RsFunction>() ?: return
+        val inserted = insertStruct(ctx.target, struct, function)
+
+        PsiDocumentManager.getInstance(project).doPostponedOperationsAndUnblockDocument(editor.document)
+
+        val fields = inserted.blockFields?.namedFieldDeclList.orEmpty()
+        if (inserted.containingFile == function.containingFile && fields.isNotEmpty()) {
+            editor.caretModel.moveToOffset(fields[0].textOffset)
+            RsMultipleVariableRenamer(fields.map { it.createSmartPointer() })
+                .doRename(fields[0], editor, DataManager.getInstance().getDataContext(editor.component))
+        } else {
+            inserted.navigate(true)
+        }
+    }
+
+    private fun buildStruct(project: Project, ctx: Context): RsStructItem? {
+        val factory = RsPsiFactory(project)
+        val fieldList = ctx.literalElement.structLiteralBody.structLiteralFieldList
+        val bindings = ctx.literalElement.getVisibleBindings()
+        val visibility = getVisibility(ctx.target, ctx.literalElement.containingMod)
+
+        val fieldsJoined = fieldList.joinToString(separator = ",\n") {
+            val name = it.referenceName
+            val expr = it.expr
+            val type = when {
+                expr != null -> expr.type
+                else -> bindings[name]?.type ?: TyUnknown
+            }
+
+            "$visibility$name: ${type.renderInsertionSafe(includeLifetimeArguments = true)}"
+        }
+        val suffix = when {
+            fieldsJoined.isEmpty() -> ";"
+            else -> " {\n$fieldsJoined\n}"
+        }
+
+        return factory.tryCreateStruct("${visibility}struct ${ctx.name}$suffix")
+    }
+
+    private fun insertStruct(targetModule: RsMod, struct: RsStructItem, sourceFunction: RsElement): RsStructItem {
+        if (targetModule == sourceFunction.containingMod) {
+            return sourceFunction.parent.addBefore(struct, sourceFunction) as RsStructItem
```

Comment:
Why do we add a struct before the function? There is a bit similar `Create class` quickfix in Kotlin and it adds classes after

Reply:
For me the logical order is structs -> functions that use them. But maybe I'm too affected by one-pass compilers (such as C/C++) :) Do you want to switch and create the struct after the function? It's true that create function intention also creates the target function after the caller.

Reply:
I feel there is some "unspoken rule" that IDEA tends to generate stuff below the place where such generation is invoked. But `structs -> functions` also sounds reasonable :thinking: 
Well, let's leave it as is for now. Nevermind